src/vuln_analysis/utils/functions_parsers/java_functions_parsers.py

```diff
@@ -27,39 +27,142 @@
+        row_index: Dict[Tuple[str, str], Dict[Tuple[str, str], int]] = {}
+        for k, lst in rows.items():
+            # stable O(1) index lookup
+            row_index[k] = {v: i for i, v in enumerate(lst)}
+
+        # 3) Upstream classifier (parents + interfaces) using relative index rule:
+        #    candidate is upstream of type_key  <=>  candidate ∈ rows[type_key],
+        #                                            type_key ∈ rows[candidate], and
+        #                                            row_index[type_key][candidate] < row_index[candidate][type_key]
+        def upstream_neighbors(type_key: Tuple[str, str]) -> List[Tuple[str, str]]:
+            """
+            Return the direct upstream types (superclasses and implemented interfaces) for `type_key`
+            using the relative-index rule over `rows` and `row_index`.
+            """
+            type_row = rows.get(type_key)
+            if not type_row:
+                return []
+
+            # Fast lookup: neighbor -> index in type_key's row
+            type_index_map = row_index.get(type_key, {})
+
+            upstreams: List[Tuple[str, str]] = []
+            seen_candidates: set[Tuple[str, str]] = set()
+
+            for candidate in type_row:
+                if candidate == type_key or candidate in seen_candidates:
+                    continue
+
+                # Index of candidate within type_key's row
+                idx_candidate_in_type = type_index_map.get(candidate, 1 << 30)
+
+                # Index of type_key within candidate's row
+                candidate_index_map = row_index.get(candidate)
+                if not candidate_index_map:
+                    continue
+                idx_type_in_candidate = candidate_index_map.get(type_key)
+                if idx_type_in_candidate is None:
+                    continue
+
+                # Ancestor/interface test via relative positions
+                if idx_candidate_in_type < idx_type_in_candidate:
+                    upstreams.append(candidate)
+                    seen_candidates.add(candidate)
+
+            return upstreams
+
+        # 4) Merge own + inherited fields per type using BFS over upstream neighbors.
+        for tkey, own_fields in per_key_fields.items():
+            merged: List[Tuple[str, str]] = []
+            have_names: set[str] = set()  # child-over-parent shadowing
+
+            # Own first (preserve order)
+            for (fname, ftype) in own_fields:
+                if fname not in have_names:
+                    merged.append((fname, ftype))
+                    have_names.add(fname)
+
+            # BFS over upstreams
+            q: deque[Tuple[str, str]] = deque(upstream_neighbors(tkey))
+            visited: set[Tuple[str, str]] = set()
+
+            while q:
+                u = q.popleft()
+                if u in visited:
+                    continue
+                visited.add(u)
+
+                uf = per_key_fields.get(u)
```

Comment:
@tmihalac Try to use a better variable name ( more reflective and descriptive one)  than `uf`